password_enhancements.module

```diff
@@ -0,0 +1,403 @@
+  if (!empty($password = $form_state->getValue('pass'))) {
+    $roles = $form_state->getValue('roles') ?? [];
+
+    /** @var \Drupal\password_enhancements\Entity\Storage\PolicyEntityStorageInterface $password_policy_storage */
+    $password_policy_storage = Drupal::entityTypeManager()->getStorage('password_enhancements_policy');
+    $policy = $password_policy_storage->loadByRoleAndPriority($roles);
+
+    $constraints = _password_enhancements_get_constraints($roles);
+    $error_messages = [];
+    $minimum_required_constraints = $policy->getMinimumRequiredConstraints();
+    $valid_count = 0;
+    foreach ($constraints as $constraint) {
+      /** @var \Drupal\password_enhancements\Plugin\PasswordConstraintPluginInterface $plugin_instance */
+      $plugin_instance = Drupal::service('plugin.manager.password_constraint')->createInstance($constraint->getType(), $constraint->getConfiguration());
+
+      if (!$plugin_instance->validate($password)) {
+        $error_messages[] = $plugin_instance->getErrorMessage();
+      }
+      else {
+        $valid_count++;
+      }
+
+      if ($minimum_required_constraints <= $valid_count) {
+        break;
+      }
+    }
+
+    // Display any validation error messages.
+    if ($error_messages !== [] && $valid_count < $minimum_required_constraints) {
+      $messages = [
+        '#theme' => 'password_enhancements_error_messages',
+        '#error_message' => t("The given password doesn't meet some requirements:"),
+        '#messages' => $error_messages,
+      ];
+      $form_state->setError($form['account']['pass'], Drupal::service('renderer')->render($messages));
+    }
+  }
+}
+
+/**
+ * Submit callback for password_enhancements_form_user_register_form_alter().
+ *
+ * This callback runs only if the password field was set to always require and
+ * the verify_email set to true.
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ *
+ * @throws \Drupal\Core\Entity\EntityStorageException
+ *   If the user was failed to save.
+ */
+function password_enhancements_form_user_register_form_alter_password_submit(array &$form, FormStateInterface $form_state): void {
+  /** @var \Drupal\user\UserInterface $user */
+  $user = $form_state->getFormObject()->getEntity();
+  $user->setPassword($form_state->getUserInput()['pass']['pass1'])
+    ->save();
+}
+
+/**
+ * Get password policy constraints by the given roles.
+ *
+ * @param array $roles
+ *   The roles for which the constraints needs to be loaded.
+ *
+ * @return \Drupal\password_enhancements\Entity\Constraint[]
+ *   Constraint list.
+ *
+ * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+ * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+ */
+function _password_enhancements_get_constraints(array $roles): array {
+  // Get entity type manager.
+  $entity_type_manager = Drupal::entityTypeManager();
```

Comment:
Unnecessary variable.

Reply:
Nope, it is being used twice in this function.

Reply:
Twice justifies the need for a new variable which is a shortcut of a shortcut? :) 
```
  public static function entityTypeManager() {
    return static::getContainer()->get('entity_type.manager');
  }
```

Reply:
I think it is generally a good idea to always put a function call's result into a variable if its results needs to be used more than one place, it maybe be a shortcut, but in another instance if the function would do something expensive, it wouldn't be a good idea to call it multiple times.

Reply:
No need, the return type of this is documented on the method, so phpstorm can recognize it's type.

Reply:
Hm, this must be a new feature. I remember that since Drupal 8.4, PHPstorm could not auto-complete methods from objects returned by `Drupal::`.